### PR TITLE
newContainer: Not attach device if it is a CDROM

### LINKED ISSUE #829: "docker run --privileged" got "docker: Error response from daemon: OCI runtime create failed: QMP command failed: unknown."
10月 17 15:28:30 teawater-virtual-machine kata-runtime[91986]: time="2018-10-17T15:28:30.891596459+08:00" level=info msg="Attaching block device" arch=amd64 command=create container=5edec94fbf4076667aa8ee546f10cde98d6ef62f3097c2a3889ee316ea0fd0e9 device=/dev/sr0 name=kata-runtime pid=91986 source=virtcontainers subsystem=device
10月 17 15:28:30 teawater-virtual-machine kata-runtime[91986]: time="2018-10-17T15:28:30.891779981+08:00" level=info msg="{\"arguments\":{\"driver\":\"raw\",\"file\":{\"driver\":\"file\",\"filename\":\"/dev/sr0\"},\"node-name\":\"drive-e6670677eccb9cab\"},\"execute\":\"blockdev-add\"}" arch=amd64 command=create container=5edec94fbf4076667aa8ee546f10cde98d6ef62f3097c2a3889ee316ea0fd0e9 name=kata-runtime pid=91986 source=virtcontainers subsystem=qmp
10月 17 15:28:30 teawater-virtual-machine kata-runtime[91986]: time="2018-10-17T15:28:30.977805252+08:00" level=info msg="{\"error\": {\"class\": \"GenericError\", \"desc\": \"Could not open '/dev/sr0': No medium found\"}}" arch=amd64 command=create container=5edec94fbf4076667aa8ee546f10cde98d6ef62f3097c2a3889ee316ea0fd0e9 name=kata-runtime pid=91986 source=virtcontainers subsystem=qmp
10月 17 15:28:30 teawater-virtual-machine kata-runtime[91986]: time="2018-10-17T15:28:30.978272497+08:00" level=info msg="Unplugging block device" arch=amd64 command=create container=5edec94fbf4076667aa8ee546f10cde98d6ef62f3097c2a3889ee316ea0fd0e9 device=/dev/dm-0 name=kata-runtime pid=91986 source=virtcontainers subsystem=device

### PR DESCRIPTION
**Pull request by teawater**
Got "docker: Error response from daemon: OCI runtime create failed:
QMP command failed: unknown." when "docker run --privileged" with kata.
In qemu part, it got:
"Could not open '/dev/sr0': Read-only file system"
or
"No medium found"
The cause is qemu need open block device to get its status.
But /dev/sr0 is a CDROM that cannot be opened.

This patch let newContainer doesn't attach device if it is a CDROM
to handle the issue.

Fixes #829

Signed-off-by: Hui Zhu <teawater@hyper.sh>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Hi @teawater - thanks for raising. I'm a little confused though - we don't support `--privileged`:

- https://github.com/kata-containers/documentation/blob/master/Limitations.md#docker-run---privileged

**Comment by lifupan** [comment]
Hi @jodh-intel, can we give a container right to access the devices in the guest using --privileged ? Since 
when add a block device to container using "-v /dev/sdbX:/dev/sdbX", kata only passthrough this
block device into container,  but it cannot be mounted since container doesn't has the right.

**Comment by raravena80** [comment]
@teawater ping, any updates?

**Comment by teawater** [comment]
I didn't see any good reason that kata doesn't support "--privileged".
Maybe it is not a safe way.  But "--privileged" is a very simple way to help the user access block device inside the container at least.  (BTW, we have a long way to go with access block device.  I think we need a configurable way.)

**Comment by amshinde** [comment]
@teawater Kata does support some notion of privileged. The container has access to all devices inside the guest, we pass as many devices as we can from the host and the container should run with all the capabilities that are associated with privileged mode, basically all capabilities.
 I am interested in understanding what is the exact error you got with qmp first, rather than ignoring the error altogether.
@lifupan You should be able to mount a device in privileged mode, please file a bug if you are seeing otherwise. I have tried this in the past simply by passing --cap-add SYS_ADMIN to the container.

**Comment by teawater** [comment]
@amshinde This is a qemu issue.  qemu need open block device to get its status.
But /dev/sr0 is a cdrom.
With disk inside, qemu will get:
"Could not open '/dev/sr0': Read-only file system\"
Without disk inside, qemu will get:
No medium found
So I think ignore the error is a suitable way to handle it.

**Comment by WeiZhang555** [comment]
Ignoring the "AttachError" may cover some real issue, so I think it's not safe.

@jodh-intel We do actually support `--privileged`, except the `/dev/sr0` issue found by @teawater , I've met same issue before :-) , without cdrom, `--privileged` can work in "most" cases :-)

@lifupan 
>Hi @jodh-intel, can we give a container right to access the devices in the guest using --privileged ? Since when add a block device to container using "-v /dev/sdbX:/dev/sdbX", kata only passthrough this
block device into container, but it cannot be mounted since container doesn't has the right.

This is a good question, we have a feature request discussing this, see 
https://github.com/kata-containers/runtime/issues/571

In a word, I think you should first considering `--cap-add SYS_ADMIN` as @amshinde suggested, `--privileged` may not be safe though it's convenient.

We'd better find a more robust way to handle `/dev/sr0` error than ignoring the `AttachError`

**Comment by teawater** [comment]
I make a patch that not attach device if it is a CDROM.

File: virtcontainers/container.go [hunk 1]
```diff
@@ -28,6 +28,23 @@ import (
 	"github.com/kata-containers/runtime/virtcontainers/utils"
 )
 
+var cDROMMajors = map[int64]string{
```
Comment (WeiZhang555) [inline virtcontainers/container.go]: The name looks weird, how about cdromMajors?

Also can you add some comments with reference links to this major definition? Then we can easily check if these are all :-)

**Comment by teawater** [comment]
@WeiZhang555 made a new version according to your comments.

**Comment by WeiZhang555** [comment]
LGTM

/test

**Comment by bergwolf** [comment]
Is CDROM the only device type we cannot handle right now? What about floppy disks? There are also other types of devices like `/dev/kvm`, `/dev/autofs` etc. that can be confusing to users if they represent the host ones. I wonder if we want to use a white list instead.

**Comment by jshachm** [comment]
+1 @bergwolf   
Now that we begin to deal with the special device, white list is better.

**Comment by teawater** [comment]
I agree with a device list.
And what about let it can be config by the config file?

**Comment by amshinde** [comment]
@bergwolf Today we ignore all devices except block and vfio.

File: virtcontainers/container.go [hunk 2]
```diff
@@ -609,12 +640,13 @@ func newContainer(sandbox *Sandbox, contConfig ContainerConfig) (*Container, err
...
 			}
 
-			c.devices = append(c.devices, ContainerDevice{
+			storedDevices = append(storedDevices, ContainerDevice{
 				ID:            dev.DeviceID(),
 				ContainerPath: info.ContainerPath,
 			})
 		}
 	}
+	c.devices = filterDevices(sandbox, c, storedDevices)
```
Comment (sboeuf) [inline virtcontainers/container.go]: The filtering should not happen every time the `container` instance is created in memory. It only needs to happen at `create`, meaning this filtering function should only be called from the case `err != nil` here.

**Comment by sboeuf** [comment]
@grahamwhaley is the failure on metrics-16-04 expected?
If that's the case, and if this is okay to ignore it, please go ahead and merge this ;)

**Comment by grahamwhaley** [comment]
Yep - metrics CI is coming back up today - just found one more 'cleanup' issue to fix now we are on baremetal build slaves. so, ignore metrics here right now, and i'll merge...

**Pull request merged**